server/webhooks/failing_policies.go

```diff
@@ -0,0 +1,131 @@
+	"github.com/go-kit/kit/log/level"
+)
+
+func TriggerFailingPoliciesWebhook(
+	ctx context.Context,
+	ds fleet.Datastore,
+	logger kitlog.Logger,
+	appConfig *fleet.AppConfig,
+	failingPoliciesSet service.FailingPolicySet,
+	now time.Time,
+) error {
+	if !appConfig.WebhookSettings.FailingPoliciesWebhook.Enable {
+		return nil
+	}
+
+	level.Debug(logger).Log("enabled", "true")
+
+	globalPoliciesURL := appConfig.WebhookSettings.FailingPoliciesWebhook.DestinationURL
+	if globalPoliciesURL == "" {
+		level.Info(logger).Log("msg", "empty global destination_url")
+		return nil
+	}
+
+	configuredPolicyIDs := make(map[uint]struct{})
+	for _, policyID := range appConfig.WebhookSettings.FailingPoliciesWebhook.PolicyIDs {
+		configuredPolicyIDs[policyID] = struct{}{}
+	}
+	policySets, err := failingPoliciesSet.ListSets()
+	if err != nil {
+		return ctxerr.Wrap(ctx, err, "listing global policies set")
+	}
+	serverURL, err := url.Parse(appConfig.ServerSettings.ServerURL)
+	if err != nil {
+		return ctxerr.Wrap(ctx, err, "invalid server url")
+	}
+	for _, policyID := range policySets {
+		if _, ok := configuredPolicyIDs[policyID]; !ok {
+			level.Debug(logger).Log("msg", "removing policy from set", "id", policyID)
+			// Remove and ignore the policies that are present in the set and
+			// are not present in the config.
+			// This could happen with:
+			//	- policies that have been deleted.
+			//	- policies with automation disabled.
+			if err := failingPoliciesSet.RemoveSet(policyID); err != nil {
+				return ctxerr.Wrapf(ctx, err, "removing global policy  %d from policy set", policyID)
+			}
+			continue
+		}
+		policy, err := ds.Policy(ctx, policyID)
+		switch {
+		case err == nil:
+			// OK
+		case errors.Is(err, sql.ErrNoRows):
+			level.Debug(logger).Log("msg", "skipping removed policy", "id", policyID)
+			// The policy might have been deleted, thus we remove the set.
+			if err := failingPoliciesSet.RemoveSet(policyID); err != nil {
+				return ctxerr.Wrapf(ctx, err, "removing global policy %d from policy set", policyID)
+			}
+			continue
+		default:
+			return ctxerr.Wrapf(ctx, err, "failing to load global failing policies set %d", policyID)
+		}
+		hosts, err := failingPoliciesSet.ListHosts(policyID)
+		if err != nil {
+			return ctxerr.Wrapf(ctx, err, "listing hosts for global failing policies set %d", policyID)
+		}
+		if len(hosts) == 0 {
+			level.Debug(logger).Log("policyID", policyID, "msg", "no hosts")
+			continue
+		}
+		failingHosts := make([]FailingHost, len(hosts))
+		for i := range hosts {
+			failingHosts[i] = makeFailingHost(hosts[i], *serverURL)
+		}
+		sort.Slice(failingHosts, func(i, j int) bool {
+			return failingHosts[i].ID < failingHosts[j].ID
+		})
+		payload := FailingPoliciesPayload{
+			Timestamp:    now,
+			Policy:       policy,
+			FailingHosts: failingHosts,
+		}
+		level.Debug(logger).Log("payload", payload, "url", globalPoliciesURL)
+		err = server.PostJSONWithTimeout(ctx, globalPoliciesURL, &payload)
```

Comment:
This function has a 30 sec timeout. That was ok for the other webhook, but this one might be a bit more active. We should reduce that to 10 secs.

Reply:
We should implement a webhook receiver that takes timeout-1 secs to respond or even timesout every time to make sure fleet doesn't drag with that kind of behavior.

Reply:
Actually we may need a bigger timeout (to be tested), e.g. ~100k failing hosts would end up with ~20MB POST body, is 30 seconds enough...?

Reply:
Good point, we could break it down into batches in that case, and do many small requests instead of one big one. The receiving end is probably not going to have more than 30 sec timeout either.

Reply:
Added a `host_batch_size` config, with `0` (no batching) by default.